core/arch/arm/kernel/ftrace.c

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <assert.h>
+#include <kernel/ftrace.h>
+#include <kernel/panic.h>
+#include <kernel/thread.h>
+#include <kernel/tee_ta_manager.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <printk.h>
+#include <string.h>
+#include <trace.h>
+#include "elf_load_dyn.h"
+
+#define MIN_FTRACE_BUF_SIZE	1024
+#define MAX_HEADER_STRLEN	128
+
+void ta_fbuf_init(vaddr_t load_addr, struct tee_ta_session *s,
+		  struct elf_load_state *state)
+{
+	struct __ftrace_info *finfo = NULL;
+	struct ftrace_buf *fbuf = NULL;
+	size_t fbuf_size = 0;
+	uintptr_t val = 0;
+	int count = 0;
+	TEE_Result res = TEE_SUCCESS;
+
+	if (state->is_32bit) {
+		DMSG("Current ftrace doesn't support 32 bit TAs");
+		return;
+	}
+
+	res = elf_resolve_symbol(state, "__ftrace_info", &val);
+	if (res != TEE_SUCCESS || !val) {
+		DMSG("ftrace info not found");
+		return;
+	}
+
+	finfo = (struct __ftrace_info *)(val + load_addr);
+	if (finfo->buf_end > finfo->buf_start)
+		fbuf_size = finfo->buf_end - finfo->buf_start;
+
+	if (fbuf_size >= MIN_FTRACE_BUF_SIZE) {
+		fbuf = (struct ftrace_buf *)finfo->buf_start;
+		fbuf->head_off = sizeof(struct ftrace_buf);
+		count = snprintk((char *)fbuf + fbuf->head_off,
+				 MAX_HEADER_STRLEN,
+				 "Function graph for TA: %pUl @ %lx\n",
+				 (void *)&s->ctx->uuid, load_addr);
+		assert(count < MAX_HEADER_STRLEN);
+
+		fbuf->ret_func_ptr = finfo->ret_ptr;
+		fbuf->ret_idx = 0;
+		fbuf->lr_idx = 0;
+		fbuf->buf_off = fbuf->head_off + count;
+		fbuf->curr_size = 0;
+		fbuf->max_size = fbuf_size - sizeof(struct ftrace_buf) - count;
+
+		s->fbuf_paddr = virt_to_phys(fbuf);
+		s->fbuf_size = fbuf_size;
+	} else {
+		DMSG("ftrace buffer too small");
+
+		s->fbuf_paddr = 0;
+		s->fbuf_size = 0;
+	}
+}
+
+void ta_fbuf_dump(struct tee_ta_session *s)
+{
+	struct ftrace_buf *fbuf = NULL;
+	struct mobj *mobj = NULL;
+	char *va = NULL;
+	uint32_t ret = 0;
+	uint32_t dump_size = 0;
+	struct thread_param params[3] = { };
+
+	if (!s->fbuf_size)
+		return;
+
+	if (!core_mmu_add_mapping(MEM_AREA_TA_RAM, s->fbuf_paddr,
```

Comment:
How about relying on the user space context being set (and thus mapped) instead. This would then work with the pager seamlessly.

How is this mapping later removed?

Reply:
> How about relying on the user space context being set (and thus mapped) instead.

This approach works during aborts where user space mappings are active. But they won't be active in normal case after close session.

> How is this mapping later removed?

`core_mmu_add_mapping()` seems to override existing mapping in reserved va_space in case it doesn't find empty reserved va_space. So we don't need to remove mapping.

Reply:
>> How about relying on the user space context being set (and thus mapped) instead.

> This approach works during aborts where user space mappings are active. But they won't be active in normal case after close session.

Given that the context still exists you can do `tee_mmu_set_ctx(ctx)` and later `tee_mmu_set_ctx(NULL)`. I guess it's best to put the responsibility on the caller to make sure that the context is mapped.
Mapping random parts physical memory without unmapping it later seems a bit shaky at best. 

>> How is this mapping later removed?

> core_mmu_add_mapping() seems to override existing mapping in reserved va_space in case it doesn't find empty reserved va_space. So we don't need to remove mapping.

￼
No, it's not overridden. It's reused if already present. I think the test cases are working because of a bit of luck and that the physical address usage is fairly static.

Reply:
> Given that the context still exists you can do tee_mmu_set_ctx(ctx) and later tee_mmu_set_ctx(NULL). I guess it's best to put the responsibility on the caller to make sure that the context is mapped.

This is something that I avoided earlier to do in kernel context after user mapping is removed. But if we can run pager with it then will use it instead.

> No, it's not overridden.

Have a look here: https://github.com/OP-TEE/optee_os/blob/ad443200b69711ad9dc30f32847b6a0d7e2bddf2/core/arch/arm/mm/core_mmu.c#L1692

Reply:
Yes?
After this has happened we can't do any more mappings in the `MEM_AREA_RES_VASPACE` range.

Reply:
> After this has happened we can't do any more mappings in the MEM_AREA_RES_VASPACE range.

It seems `MEM_AREA_RES_VASPACE` map entry will be overridden again and again [1] in case size of buffer is less than `map->size` after filling in empty `static_memory_map` map entries.

[1] Else region indicates as follows:

```
        } else {
                /*
                 * There isn't room for another entry, steal the reserved
                 * entry as it's not useful for anything else any longer.
                 */
                map->size = l;
        }
```

Reply:
`MEM_AREA_RES_VASPACE` is gone after this with the `map->type = type` assignment below. So it's not possible to map anything in that range any longer.

Reply:
Ah, I see. I almost overlooked it.

Reply:
It seems that I can't directly use `tee_mmu_set_ctx(ctx)` and `tee_mmu_set_ctx(NULL)` as it fails during inter user TA communication. So need to use `tee_ta_push_current_session()` and `tee_ta_pop_current_session()` apis.

Reply:
OK, if the session is in a usable state that's fine.